ERCS/erc-7683.md

```diff
@@ -72,6 +72,100 @@ Cross-chain execution systems implementing this standard SHOULD use a sub-type t
+
+// The Message contains calls that the user wants executed on the destination chain.
+// The target is a contract on the destination chain that the settlement contract will attempt to send callData and value to.
+struct Calls {
+  address target;
+  bytes callData;
+  uint256 value;
+}
+
+// A fallbackRecipient can be set by the user in case sending the callData to the target reverts and the user still wants to guarantee that they receive their value to an EOA
+// on the destination chain.
+struct Message {
+  Calls[] calls;
+  address fallbackRecipient;
+}
+```
+
+The `Message` sub-type is designed to be used by a 7683 user to incentivize a filler to to execute arbitrary calldata on a target destination chain contracton the user's behalf. For example, the settlement contract might decode the `orderData` containing the message information as follows:
+
+```solidity
+function fill(bytes32 orderId, bytes calldata originData, bytes calldata fillerData) public {
+	// Input[] userInputs;
+	// Output[] maxOutputs;
+	// Output[] minFillerOutputs;
+	(
+		address user,
+		uint32 originChainId,
+		uint32 initiateDeadline,
+		uint32 fillDeadline,
+		Input[] memory userInputs,
+		Output[] memory maxOutputs,
+ 		Output[] memory minFillerOutputs,
+		Instructions memory instructions
+	) = abi.decode(originData);
+
+	// ...Do some preprocessing on the parameters here to validate a ResolvedCrossChainOrder...
+
+	// ...Execute the fill logic of the ResolvedCrossChainOrder...
+
+	// Handle the Message subtype:
+
+        // If there is no fallback recipient, call and revert if the inner call fails.
+        if (instructions.fallbackRecipient == address(0)) {
+            this.attemptCalls(instructions.calls);
+            return;
+        }
+
+        // Otherwise, try the call and send to the fallback recipient if any tokens are leftover.
+        (bool success, ) = address(this).call(abi.encodeCall(this.attemptCalls, (instructions.calls)));
+        if (!success) emit CallsFailed(instructions.calls, instructions.fallbackRecipient);
+
+        // If there are leftover tokens, send them to the fallback recipient regardless of execution success.
+        _drainRemainingTokens(token, payable(instructions.fallbackRecipient));
+    }
+
+    function attemptCalls(Call[] memory calls) external onlySelf {
+        uint256 length = calls.length;
+        for (uint256 i = 0; i < length; ++i) {
+            Call memory call = calls[i];
+
+            // If we are calling an EOA with calldata, assume target was incorrectly specified and revert.
+            if (call.callData.length > 0 && call.target.code.length == 0) {
+                revert InvalidCall(i, calls);
+            }
+
+            (bool success, ) = call.target.call{ value: call.value }(call.callData);
+            if (!success) revert CallReverted(i, calls);
+        }
+    }
+
+    function _drainRemainingTokens(address token, address payable destination) internal {
```

Comment:
While I think this is closer to what a user/integrator might really want, the complexity may be intimidating in a canonical example. Thoughts on showing something simpler, like one where if the calls fail, the entire fill reverts with no fallback?

Reply:
Ok yeah but we should add the drain logic to the erc7683 repo if we want to seed it with a subtype